File: 2017/day1/part2.rb
sum = 0
nums = File.read("input.txt").chomp.chars.map(&:to_i)

halves = nums.each_slice(nums.length / 2).to_a
pairs = halves[0].zip(halves[1])
sum = pairs.reduce(0) do |acc, pair|
  acc += pair[0] == pair[1] ? pair[0] * 2 : 0
end
puts sum

File: 2017/day1/part2.exs
nums = File.read!("input.txt")
|> String.trim
|> String.graphemes
|> Enum.map(&String.to_integer/1)

{first, second} = Enum.split(nums, div(Enum.count(nums),2))

Enum.zip(first, second)
|> Enum.filter(fn{x,y} -> x == y end)
|> Enum.map(fn{x,_} -> x * 2 end)
|> Enum.sum
|> IO.inspect

# Improvements
# -------
# There is a `match` function which can pattern match which would make 9 much simpler
#   => Enum.filter(&match?({x, x}, &1))
# There is an `elem` function for grabbing an element out of a tuple for line 10
#   => Enum.map(&elem(&1, 0))
